### DISCUSSION TITLE
ASK 피처 개발 실전 돌입하기!!

### DISCUSSION THREAD

**Discussion by jopopcorn** [euphony-io/euphony#118] [General]
혼자서 최대한 짜보고 성호님과도 상의를 해봤는데 c++을 모르는 사람들끼리 완벽하게 해내기에는 한계가 있는거같아 디스커션을 올립니다 🥲
평가 기간이 하루도 남지 않은 시간인데 끙끙대다가 뒤늦게 올려서 도움 주시는 멘토님들, 멘티분들께 정말 죄송스럽네요.. 😭
제가 짠 ASK 관련 코드를 리뷰해주시면 정말 감사할 것 같습니다!!

</br>

## ASK.h
모뎀 인터페이스를 상속받아서 다음과 같은 메서드들로 구성해보았습니다.
```c++
#ifndef EUPHONY_ASK_H
#define EUPHONY_ASK_H

#include "FFTModel.h"
#include "Modem.h"
#include <string>
#include <vector>

namespace Euphony {
    class ASK : public Modem{
    public:
        ASK();
        WaveList modulate(std::string code);
        WaveList modulate(Packet* packet);
        std::shared_ptr<Packet> demodulate(const WaveList& waveList);
        std::shared_ptr<Packet> demodulate(const float* source, int sourceLength, int bufferSize);

    private:
        std::unique_ptr<FFTModel> fftModel;
    };
}
#endif //EUPHONY_ASK_H

```

</br>

## ASK.cpp
demodulate는 아직 구현하지 않았고 FSK 코드를 임의로 붙여넣은 것이므로 무시해주세요!
demodulate는 [제가 생각한 modulate 방식](https://github.com/euphony-io/euphony/discussions/85#discussioncomment-1455043)의 반대로 구현하려고 하는데 혹시 좋은 방법이 있다면 조언 부탁드립니다!! 🥺
지웅 멘토님께서 [구현에 관한 조언](https://github.com/euphony-io/euphony/discussions/115#discussioncomment-1459348)을 해주셨는데 이것도 참고하면 좋을 것 같습니다!
modulate 코드에서 0과 1을 표현하려면 `vibratesAt` 값에 무엇을 넣어줘야 적절할지 확실하지 않는데, 이부분도 같이 고민해주시면 좋을 것 같아요!
```c++
#include "../ASK.h"
#include "../FFTProcessor.h"
#include "../Definitions.h"
#include "../WaveBuilder.h"
#include "../Base16Exception.h"
#include "../Packet.h"
#include <cmath>

using namespace Euphony;

ASK::ASK() : fftModel(std::make_unique<FFTProcessor>(kFFTSize, kSampleRate)) {}

WaveList ASK::modulate(Packet *packet) {
    return this->modulate(packet->toString());
}

WaveList ASK::modulate(std::string code) {
    vector<shared_ptr<Wave>> result;

    for (char c: code) {
        switch (c) {
            case '0':
                result.push_back(
                        Wave::create()
                                .vibratesAt(0)
                                .setSize(kBufferSize)
                                .setCrossfade(BOTH)
                                .build()
                );
                break;
            case '1':
                result.push_back(
                        Wave::create()
                                .vibratesAt(kStandardFrequency)
                                .setSize(kBufferSize)
                                .setCrossfade(BOTH)
                                .build()
                );
                break;
            case 'S':
                result.push_back(
                        Wave::create()
                                .vibratesAt(kStartSignalFrequency)
                                .setSize(kBufferSize)
                                .setCrossfade(BOTH)
                                .build()
                );
                break;
        }
    }

    return result;
}

std::shared_ptr<Packet> ASK::demodulate(const WaveList &waveList) {
    HexVector hexVector = HexVector(waveList.size());

    for (const auto &wave : waveList) {
        auto vectorInt2Source = wave->getInt16Source();
        int16_t *int16Source = &vectorInt2Source[0];
        float *resultBuf = fftModel->makeSpectrum(int16Source);
//        hexVector.pushBack();
    }
    return std::make_shared<Packet>(BaseType::BASE2, hexVector);
}

std::shared_ptr<Packet>
ASK::demodulate(const float *source, int sourceLength, int bufferSize) {
    int dataSize = sourceLength / bufferSize;
    HexVector hexVector = HexVector(dataSize);

    WaveList waveList;
    for(int i=0; i<dataSize; i++){
        waveList.push_back(std::make_shared<Wave>(source + (i * bufferSize), bufferSize));
    }
    return demodulate(waveList);
}

```

</br>

## ASKTest.cpp
우선 테스트코드는 작성해보았는데 modulate를 검증하려면 테스트 케이스를 어떻게 작성해서 확인해야할지 모르겠습니다.. 😿
제가 원하는 결과는 "modulate로 변환한 값과 param에 입력한 값이 같다" 인데요.
`expectedCodeLength`랑 같지 않는 이유는 `modulateASK.size()`의 값이 패킷까지 포함해서 나와서 사이즈가 길다는 추측입니다..
이걸 빌드해봤더니 모두 fail이 나오더라고요.. 여기서 어떻게 고쳐야 좋을지 막막해서 일단 그대로 올려봅니다
```c++
#include <gtest/gtest.h>
#include <Definitions.h>
#include <ASK.h>
#include <FFTProcessor.h>
#include <tuple>
#include <ASCIICharset.h>
#include <HexVector.h>
#include <Packet.h>

using namespace Euphony;

typedef std::tuple<std::string, int> TestParamType;

class ASKTestFixture : public ::testing::TestWithParam<TestParamType>{

public:
    void createASK() {
        EXPECT_EQ(ask, nullptr);
        ask = new ASK();
        ASSERT_NE(ask, nullptr);
    }

    ASK* ask = nullptr;
    FFTProcessor* fft = nullptr;
};

TEST_P(ASKTestFixture, ASKModulationStringTest){
    createASK();

    string source;
    string expectedModulateResult;
    int expectedCodeLength;

    std::tie(source, expectedCodeLength) = GetParam();
    HexVector hexVector = ASCIICharset().encode(source);
    auto *packet = new Packet(BaseType::BASE2, hexVector);

    auto modulateASK = ask->modulate(packet->toString());
    EXPECT_EQ(modulateASK.size(), expectedCodeLength);

//    auto demodulateResult = ask->demodulate(modulateASK);
//    EXPECT_EQ(source, demodulateResult->getPayloadStr());
}

INSTANTIATE_TEST_CASE_P(
        ASKTestSuite,
        ASKTestFixture,
        ::testing::Values(
                TestParamType("0", 1),
                TestParamType("1", 1),
                TestParamType("2", 1),
                TestParamType("3", 1),
                TestParamType("4", 1),
                TestParamType("5", 1),
                TestParamType("6", 1),
                TestParamType("7", 1),
                TestParamType("8", 1),
                TestParamType("9", 1),
                TestParamType("10", 2)
                )
        );
```

</br></br>

이 외에도 빌드를 위해서는 여러가지 파일을 건드려야 합니다!! 이것도 추가로 올리는게 좋을 것 같아서 부가적으로 설명 추가합니다!

</br>

## Definitions.h
유포니에서 쓰이는 글로벌 상수들 중 ModulationType(36 line)에 ASK 주석처리 해제하고 다음과 같이 추가해주세요.
```c++
    enum class ModulationType : int32_t {
        FSK = 0,

        // TODO: v0.7.1.6 had ASK feature. but v0.8 has to create it.
        ASK = 1,

        /*
        TODO: Rearchitecturing necessary because the CPFSK modulation type has some glitch sound.
        CPFSK
         */
    };
```

</br>

## ModemFactory.h
모뎀 팩토리에 ASK를 추가해주세요.
```c++
#ifndef EUPHONY_MODEMFACTORY_H
#define EUPHONY_MODEMFACTORY_H

#include "Definitions.h"
#include "Modem.h"
#include "FSK.h"
#include "ASK.h"

namespace Euphony {
    class ModemFactory {
    public:
        static std::shared_ptr<Modem> create(ModulationType type) {
            switch(type) {
                default:
                case ModulationType::FSK :
                    return std::make_shared<FSK>();
                case ModulationType::ASK :
                    return std::make_shared<ASK>();
            }
        }
    };
}

#endif //EUPHONY_MODEMFACTORY_H

```

</br>

### cpp/tests/CMakeLists.txt
11 line의 add_executable 함수에 ASK Test 파일을 추가해줍니다.
```c++
add_executable(
        ${TEST_EUPHONY}
        ASKTest.cpp
        ...
```

### cpp/CMakeLists.txt
24 line의 유포니 소스코드를 세팅하는 명령어에 ASK 파일을 추가해줍니다.
```c++
set(EUPHONY_SRC
        arms/kiss_fft.c
        arms/kiss_fftr.c
        core/source/AudioStreamCallback.cpp
        core/source/ASCIICharset.cpp
        core/source/ASK.cpp
        ...
```

</br>

코드를 보고 질문이나 조언, 해주고 싶은 말이 있다면 정말 자유롭게 코멘트 달아주시면 감사하겠습니다!!

**Comment by designe** [comment]
오 거의다 하셨군요!
유닛테스트 정답 부분이 수정되야하지 않을까 합니다
단계별로 확인해보는게 좋아요!

```cpp
    std::tie(source, expectedCodeLength) = GetParam();
    HexVector hexVector = ASCIICharset().encode(source);
    auto *packet = new Packet(BaseType::BASE2, hexVector);

    auto modulateASK = ask->modulate(packet->toString());
    EXPECT_EQ(modulateASK.size(), expectedCodeLength);
```

작성하신 코드에서 packet을 통해 예상하는 결과 값은?
https://github.com/euphony-io/euphony/blob/ee76f7890caf24848d09a3109b7b3be745d82ac5/euphony/src/main/cpp/tests/packetTest.cpp#L31
https://github.com/euphony-io/euphony/blob/master/euphony/src/main/cpp/tests/base2Test.cpp

위 2개의 유닛테스트 결과값을 통해 유추할 수 있을 것 같아요!
이진수라서 꽤 길거에요!

**Reply by jopopcorn** [reply]
패킷과 Base2 테스트코드에 좋은 단서가 있었군요!!
감사합니다! 참고해서 다시 작성해보겠습니다!

**Comment by kuro11pow2** [comment]
와 많은 진전이 있었네요! 저도 한번 코드를 적용해보겠습니다! 👍

**Reply by jopopcorn** [reply]
앗 혹시나 필요하신 분 있을까봐 origin에 브랜치 파서 올려놓을까 했는데 지금까지 작업한 코드 올려놓겠습니다!
여기에 올려놓았으니 다른 분들은 가져와서 작업하시면 될 것 같습니다!
https://github.com/jopopcorn/euphony/tree/feature/ask

**Comment by kuro11pow2** [comment]
감사합니다!
저희는 TDD로 개발하니까 궁극적으로 해야하는 일이 ASKTest.cpp를 통과시키는 거라고 생각합니다!
그러니까 ASKTest.cpp를 먼저 작성하고 유닛테스트까지 정한 후에 ASK 클래스를 작성해보면 어떨까 싶네요 

```cpp
    std::tie(source, expectedCodeLength) = GetParam();
    HexVector hexVector = ASCIICharset().encode(source);
    auto *packet = new Packet(BaseType::BASE2, hexVector);

    auto modulateASK = ask->modulate(packet->toString());
    EXPECT_EQ(modulateASK.size(), expectedCodeLength);
```

이 부분을 보면 
변환할 데이터, 예상 코드 길이를 받아서
데이터를 아스키로 인코딩하고, Base2인 패킷으로 만듭니다. 
패킷을 문자열로 바꾸고 modulate해서 그 길이가 예상 코드 길이와 같은지 봅니다.

"0"는 ascii로 0x30 입니다.
Base2로 하면 0b00110000 입니다.
그럼 패킷은 S + 00110000 + checksum의 문자열 + parityCode의 문자열 이 되겠죠

PacketErrorDetector.cpp을 살펴보면 checksum과 parity는 각각 4비트입니다. 
그럼 전체 길이는 1 + 8 + 4 + 4 = 17이 되겠네요

```cpp
INSTANTIATE_TEST_CASE_P(
        ASKTestSuite,
        ASKTestFixture,
        ::testing::Values(
                TestParamType("0", 17)
));
```
이렇게 만들어보면 어떨까 싶습니다.

**Reply by jopopcorn** [reply]
아하 감사합니다!! 패킷 생성할 때 checksum과 parity code의 값을 만들어주는 코드가 아직 이해되지 않아서 살펴보고 있는 중이었습니다 ㅠ
준규님께서 주신 테스트 케이스로 빌드해보았을 때 FAILED가 뜨는데 그럼 ASK 코드를 바꾸는 게 좋을까요??

**Reply by kuro11pow2** [reply]
빠트린 부분이 있어서 추가합니다!!
modulate된 결과는 wave의 벡터고. 1글자당 1 wave니까 패킷의 문자열 크기와 wave 벡터의 크기는 같습니다.
테스트 케이스 FAILED 부분은 조금 더 알아보고 다시 말씀드리겠습니다!!

**Reply by kuro11pow2** [reply]
생각해보니 checksum과 parity도 4비트씩 들어가야 하네요! "0"이면 17이 맞습니다!

**Reply by kuro11pow2** [reply]
"00" 이면 1 + 8 * 2 + 4 + 4 = 25겠네요

**Reply by kuro11pow2** [reply]
![image](https://user-images.githubusercontent.com/47289893/136911237-952c5166-3534-45db-83fb-41880f13c87e.png)

**Reply by jopopcorn** [reply]
아하 그렇군요 그런데 패킷 생성하는 과정이 아직 확실하게 이해가지 않아서 질문 드립니다 😢
제가 PacketErrorDetector의 makeChecksum 함수의 코드가 이해되지 않아서 그러는데, 혹시 이부분이 무엇을 의미하는지 설명해주실 수 있을까요??
```c++
    payloadSum &= 0xF;
    payloadSum = (~payloadSum + 1) & 0xF;
```

테스트 케이스를 작성하고 싶어도 잘 이해가 안가서 진전이 없네요 ㅠㅠ

**Reply by kuro11pow2** [reply]
(~payloadSum + 1) 이 부분은 2의 보수를 취하는 부분인데요.
예를 들어 1011 이라면 0100 + 1 = 0101이 되고 두 수를 다시 더해주면 오버플로가 발생해서 0000이 됩니다. 
보통 간단한 체크섬은 이렇게 n바이트의 데이터를 다 더하고 아래 몇 비트를 잘라서 2의 보수를 취해 만듭니다.

https://ko.wikipedia.org/wiki/%EC%B2%B4%ED%81%AC%EC%84%AC

여기 참고하시면 잘 이해하실 수 있을 것 같습니다!!

**Reply by jopopcorn** [reply]
> (~payloadSum + 1) 이 부분은 2의 보수를 취하는 부분인데요. 예를 들어 1011 이라면 0100 + 1 = 0101이 되고 두 수를 다시 더해주면 오버플로가 발생해서 0000이 됩니다.

아아아 이해했습니다!! 예전에 네트워크 배운 내용이 갑자기 생각났네요.. 감사드려요!!

**Reply by psh2849** [reply]
준규님 덕분에 테스트 케이스 몇 개 작성해봤습니다.
```c++
TestParamType("0", 17),
TestParamType("a", 17),
TestParamType("abc",33),
TestParamType("000",33),
TestParamType("0000",41),
TestParamType("000000000000000000",153),
TestParamType("abcdefghhijklmnopqrstuvwxyz", 225)                
```

**Reply by designe** [reply]
Goooooooooooooood :)

**Comment by jopopcorn** [comment]
@kuro11pow2 
알려주신 덕분에 다른 테스트 케이스도 작성해보았습니다!
그런데 다른 것들은 다 성공했는데 한글을 입력할 때만 실패가 뜹니다. 성호님이 확인해보신 결과로는 한글로 똑같이 작성해보아도 fail이 뜬다고 하는데, 한글을 아스키 코드로 변환하고 패킷을 만드는 것이 일반 케이스와 어떤 차이점이 있는건가요?? 이부분은 어떻게 케이스를 작성해줘야 할지 모르겠어서 코멘트 올립니다!

```c++
INSTANTIATE_TEST_CASE_P(
        ASKTestSuite,
        ASKTestFixture,
        ::testing::Values(
                TestParamType("0", 17),
                TestParamType("a", 17),
                TestParamType("00", 25),
                TestParamType("abcdefghijklmnopqrstuvwxyz", 217),
                TestParamType("abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyz", 425),
                TestParamType("abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyz", 633),
                TestParamType("!", 17),
                TestParamType("ㄱ", 17),
                TestParamType("가", 17),
                TestParamType("각", 17),
                TestParamType("깎", 17)
                )
        );
```

**Reply by designe** [reply]
한글은 ASCIICharset을 쓰시면 안됩니다 ㅎㅎ ASCII code 중에 한글이 없기 때문이지요.
이번에 유랑팀에서 만든 UTF8Charset 또는 UTF16Charset, UTF32Charset을 이용하시면 됩니다.

**Reply by designe** [reply]
ASCII와 호환되는게 UTF-8이라 utf8charset을 이용하는게 제일 편할 것 같군요.

**Reply by designe** [reply]
utf8에서는 한글은 3바이트로 해석됩니다. 이것도 참고하시면 유닛테스트 정답셋 만드실때 도움될 것 같습니다.

**Reply by jopopcorn** [reply]
아하 알겠습니다!! 조언 감사드려요!

**Comment by kuro11pow2** [comment]
@designe 

혹시 아래 코드에서 "S"를 어떻게 복원하시는지 설명해주실 수 있을까요?
ASK의 demodulate를 만들고 있는데 쉽지가 않네요....


FSK.cpp
```cpp
shared_ptr<Packet> FSK::demodulate(const WaveList& waveList) {
    HexVector hexVector = HexVector(waveList.size());
    
    for(const auto& wave : waveList) {
        auto vectorInt16Source = wave->getInt16Source();
        int16_t* int16Source = &vectorInt16Source[0];
        float *resultBuf = fftModel->makeSpectrum(int16Source);
        hexVector.pushBack(getMaxIdxFromSource(resultBuf));
    }

    return std::make_shared<Packet>(hexVector);
}

std::shared_ptr<Packet> FSK::demodulate(const float *source, int sourceLength, int bufferSize) {
    int dataSize = sourceLength / bufferSize;
    HexVector hexVector = HexVector(dataSize);

    WaveList waveList;
    for(int i = 0; i < dataSize; i++) {
        waveList.push_back(std::make_shared<Wave>(source + (i * bufferSize), bufferSize));
    }

    return demodulate(waveList);
}
```

**Reply by designe** [reply]
@kuro11pow2 
https://github.com/euphony-io/euphony/blob/0336d29ba1df475421554d3edb0cfd01b86989ed/euphony/src/main/cpp/tests/FSKTest.cpp#L61
```cpp
INSTANTIATE_TEST_CASE_P(
        FSKTestSuite,
        FSKTestFixture,
        ::testing::Values(
                TestParamType("0", 1, 0),
                TestParamType("1", 1, 1),
                TestParamType("2", 1, 2),
                TestParamType("3", 1, 3),
                TestParamType("4", 1, 4),
                TestParamType("5", 1, 5),
                TestParamType("012345", 6, 0),
                TestParamType("0123456789", 10, 0),
                TestParamType("abcdef", 6, 10),
                TestParamType("0123456789abcdef", 16, 0)
));
```
위는 FSK 유닛테스트 코드입니다.
payload(패킷 주 내용)와 뒤의 에러코드만 포함해서 해석하고 있습니다.
S는 FSK 에서도 해석하고 있지 않습니다. demodulate할때 S는 데이터에 포함되지 않아서 따로 생각하지 않으셔도 될 것 같습니다.
ASK 또한 S는 따로 유닛테스트에서 생각하지 않으셔도 될 것 같습니다! (페이로드 및 error 체크 데이터만 확인하시면 되겠습니다)

**Reply by designe** [reply]
FSK와 ASK의 큰 차이점이라면 ASK는 아무것도 감지되지 않을때 0을 뱉는다는 점인데요.

demodulate의 경우, FFT를 사용하고 있고, 
FFT에 대한 충분한 이해도가 없는 상태에서 구현하는 것은 어려울 수 있을 것 같아요.
사실 구현 자체는 특정 threshold를 정해서 
fftModel->makeSpectrum의 결과가 18000 이후 주파수들의 세기(amplitude)를 의미하는데 이것이 특정 threshold보다 높으면 1, 낮으면 0
으로 처리하면 됩니다만! 이 특정 주파수의 세기가 하이퍼 파라메터와 같은 사람이 정해야하는 수치입니다 ㅠㅋㅋ

https://github.com/euphony-io/euphony/blob/0336d29ba1df475421554d3edb0cfd01b86989ed/euphony/src/main/java/co/euphony/rx/EuRxManager.java#L265
여기서부터는 listener의 영역인데요.
0.0009를 threshold로 사용하고 있는데 이걸 그대로 사용해도 괜찮긴 할 것 같습니다.

**Reply by kuro11pow2** [reply]
이런 식으로 짜봤었는데 demodulate해서 구한 패킷이 이상하게 나오더라구요 ㅠㅠ

std::tie(inputCode, expectedCodeLength) = GetParam();
HexVector hexVector = ASCIICharset().encode(inputCode);
auto *packet = new Packet(BaseType::BASE2, hexVector);
packet->toString():
S0011000011010011

auto demodulateResult = ask->demodulate(modulateASK);
demodulateResult->toString():
S1001100001101001180

```cpp
std::shared_ptr<Packet> ASK::demodulate(const WaveList &waveList) {
    HexVector hexVector = HexVector(waveList.size());

    for(const auto& wave : waveList) {
        auto vectorInt16Source = wave->getInt16Source();
        int16_t* int16Source = &vectorInt16Source[0];
        float *resultBuf = fftModel->makeSpectrum(int16Source);
        const int startIdx = std::lround((static_cast<float>(static_cast<float>(kStartSignalFrequency) / static_cast<float>(kSampleRate >> 1)) * (kFFTSize >> 1)));
        const int targetIdx = getStartFreqIdx();
        if (resultBuf[startIdx] > 0.9) {
            hexVector.pushBack(startIdx - targetIdx);
        }
        else if (resultBuf[targetIdx] > 0.1) {
            hexVector.pushBack(1);
        }
        else {
            hexVector.pushBack(0);
        }
    }
    return std::make_shared<Packet>(BaseType::BASE16, hexVector);
}
```

**Reply by designe** [reply]
아래는 ASK demodulate에 대한 슈도코드인데요.

```cpp
shared_ptr<Packet> ASK::demodulate(const WaveList& waveList) {
    HexVector hexVector = HexVector(waveList.size());
    const float threshold = 0.0009;
    const int startIdx = getStartFreqIdx(sampleRate, fftSize); // fsk에서 사용되는 것과 동일

    for(const auto& wave : waveList) {
        auto vectorInt16Source = wave->getInt16Source();
        int16_t* int16Source = &vectorInt16Source[0];
        float *resultBuf = fftModel->makeSpectrum(int16Source);
        if(resultBuf[startIdx] > threshold)
            hexVector.pushBack(1);
        else
            hexVector.pushBack(0);
    }

    return std::make_shared<Packet>(hexVector);
}
```

위에 제가 쓴걸 코드로 구현해보았는데 컴파일을 해보지 않아 이런 느낌의 로직 정도로만 이해하시면 될 것 같아요.

에효 ㅋㅋ 다들 마지막날 아무쪼록 고생하시네요 ㅠㅋㅋ

**Reply by kuro11pow2** [reply]
ㅠㅠ.. 그러게요 멘토님 정말 고생하십니다 슈도코드를 베이스로 한번 만들어보겠습니다!!!

**Reply by designe** [reply]
@kuro11pow2 
demodulateResult->getPayloadStr() 
페이로드는 제대로 나오시나요?

**Reply by designe** [reply]
payload만 체크하시면 됩니다. 어짜피 이곳에서 해야할 일은 페이로드가 정상적으로 인식되냐이기 떄문에 ㅎㅎ

**Reply by designe** [reply]
S 00110000 1101 0011
S1 00110000 1101 0011 80

차이를 보니 거의 다 오신 것 같긴 하네요! 유후

**Reply by HACore** [reply]
멘토님 슈도코드를 참고해서 작성해보니 유닛테스트는 통과됐는데 아직 테스트코드가 미흡해서 제대로 구현된건지 걱정되네요😂

```cpp
std::shared_ptr<Packet> ASK::demodulate(const WaveList &waveList) {
    HexVector hexVector = HexVector(waveList.size());
    const float threshold = 0.0009;
    const int startIdx = getStartFreqIdx();

    for (const auto &wave : waveList) {
        auto vectorInt2Source = wave->getInt16Source();
        int16_t *int16Source = &vectorInt2Source[0];
        float *resultBuf = fftModel->makeSpectrum(int16Source);
        if(resultBuf[startIdx] > threshold)
            hexVector.pushBack(1);
        else
            hexVector.pushBack(0);
    }
    return std::make_shared<Packet>(hexVector);
}

int ASK::getStartFreqIdx() {
    return std::lround((static_cast<float>(static_cast<float>(kStandardFrequency) / static_cast<float>(kSampleRate >> 1)) * (kFFTSize >> 1)));
}
```

지금 테스트중인 코드는 아래와 같습니다..
곧 한글 케이스에 대한 코드도 추가해서 돌려볼게요!

```cpp
INSTANTIATE_TEST_CASE_P(
        ASKTestSuite,
        ASKTestFixture,
        ::testing::Values(
                TestParamType("0", 17),
                TestParamType("a", 17),
                TestParamType("00", 25),
                TestParamType("abcdefghijklmnopqrstuvwxyz", 217),
                TestParamType("abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyz", 425),
                TestParamType("abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyz", 633)
        )
);
```

**Reply by HACore** [reply]
> S는 FSK 에서도 해석하고 있지 않습니다. demodulate할때 S는 데이터에 포함되지 않아서 따로 생각하지 않으셔도 될 것 같습니다.
ASK 또한 S는 따로 유닛테스트에서 생각하지 않으셔도 될 것 같습니다! (페이로드 및 error 체크 데이터만 확인하시면 되겠습니다)

@designe 
modulate할때 S도 복조화해서 전송하는데 demodulate에서는 데이터에 포함되지 않는다는 부분이 모호한데,
들어오는 데이터에서는 S가 왜 포함되지 않는지 조금 더 설명 부탁드려도 될까요?

질문을 너무 많이 드린것같아 죄송합니다ㅠㅠ
해당 부분이 이해되지 않아서 현재 테스트코드 작성하는데에 난항을 겪고있네요..

**Reply by kuro11pow2** [reply]
혹시 저희가 TestParamType에 16진수가 아닌 걸 넣어서 문제가 생긴 게 아닐까요? Packet의 payload는 Base2나 Base16라서 0~f가 아닌 문자를 넣으면 예외 에러가 발생하더라구요 ㅠㅠ 0과 1만 넣어서 테스트해보니까 modulate도 잘 되고 demodulate도 잘 되는 것 같습니다!

![image](https://user-images.githubusercontent.com/47289893/136959197-196ebea2-de1c-448d-9213-0c0a8ac15760.png)

**Reply by designe** [reply]
> > S는 FSK 에서도 해석하고 있지 않습니다. demodulate할때 S는 데이터에 포함되지 않아서 따로 생각하지 않으셔도 될 것 같습니다.
> > ASK 또한 S는 따로 유닛테스트에서 생각하지 않으셔도 될 것 같습니다! (페이로드 및 error 체크 데이터만 확인하시면 되겠습니다)
> 
> @designe modulate할때 S도 복조화해서 전송하는데 demodulate에서는 데이터에 포함되지 않는다는 부분이 모호한데, 들어오는 데이터에서는 S가 왜 포함되지 않는지 조금 더 설명 부탁드려도 될까요?
> 
> 질문을 너무 많이 드린것같아 죄송합니다ㅠㅠ 해당 부분이 이해되지 않아서 현재 테스트코드 작성하는데에 난항을 겪고있네요..

@HACore 
지금에서야 확인해서 답변드려요! 
질문 하나밖에 안했는데 죄송하다니요 ㅋㅋ (오랜만에 질문을 주셨어요 ㅠㅎㅎ)

충분히 모호할 수 있습니다. 
**modulate과 demodulate는 정반대라서 깔끔하게 정반대가 되는 것이 맞는 것 아닌가요?**
라는 질문이죠?

`demodulate`는 rx 부분에서 사용되는 함수이고 `S`의 존재는 음파 인식 트리거를 위한 기준점이라고 볼 수 있습니다. 
AI Assistant로 예를 들자면 갤럭시는 하이 빅스비! Google Assistant는 Hey Google, 아이폰은 시리야 정도로 볼 수 있어요. 
이러한 특정 트리거 이후에 음성 인식 모드에 진입하게 되죠.

음파 통신은 `S`에 해당하는 (18002 - 86)hz 를 의미하고 이 주파수가 감지된 이후에 데이터를 인식합니다.
그 데이터가 `S` 다음에 오는 데이터들을 의미합니다. 위에 언급된 AI Assistant들처럼 음파 인식 단에서 listen 호출 이후로
지속적으로 `S`에 해당하는 주파수를 확인합니다. 발생하는지 안하는지 그리고 발생과 동시에
뒤에 발생하는 데이터들을 쭈우우욱 받아서 `demodulate`를 하게 될거에요. 
따라서 Modem의 `demodulate`에서 `S`에 대한 해석을 하지 않게 되었답니다.

근데 `modulate`에는 왜 `S`가 있는건가?
이것은 개발 용이성을 위해 이곳에서 구현하게 되었어요. 
주파수를 oscillate 하는 Responsibility를 Wave에서만 주고 싶었던 것이죠.

더 나은 개발 용이한 포인트를 찾게 된다면 나중에 바뀔수도 있습니다!
답변이 되었을까용?!

**Reply by designe** [reply]
> 혹시 저희가 TestParamType에 16진수가 아닌 걸 넣어서 문제가 생긴 게 아닐까요? Packet의 payload는 Base2나 Base16라서 0~f가 아닌 문자를 넣으면 예외 에러가 발생하더라구요 ㅠㅠ 0과 1만 넣어서 테스트해보니까 modulate도 잘 되고 demodulate도 잘 되는 것 같습니다!

@kuro11pow2 
엇? ASCII나 한글도 테스트하신다고 했던거 같은데 Charset은 잘 안되던가요?! 궁금하네요.

**Reply by kuro11pow2** [reply]
@designe 

ASK 개발하려구 @jopopcorn @psh2849 @HACore 님과 google meet으로 만나서 의논해봤는데요 FSK와 동일한 형태가 되도록 ASK를 만들어보려고 했습니다 그런데 그렇게 하려면 inputCode와 demodulateResult->getPayloadStr()를 비교해야 하는데, getPayloadStr은 Base 객체의 getBaseString의 결과를 return하더라구요 ㅠㅠ 다시보니까 FSKTest.cpp의 테스트 케이스도 모두 0~f 로 이루어져있더라구요. 그래서 0과 1로 구성된 문자열로만 테스트를 하게 되었습니다 😢 

https://github.com/euphony-io/euphony/blob/c74b9931b04120693f1230e27920de55b831adfb/euphony/src/main/cpp/tests/FSKTest.cpp#L32-L47

https://github.com/euphony-io/euphony/blob/c74b9931b04120693f1230e27920de55b831adfb/euphony/src/main/cpp/tests/FSKTest.cpp#L61-L75

**Reply by designe** [reply]
> @designe
> 
> ASK 개발하려구 @jopopcorn @psh2849 @HACore 님과 google meet으로 만나서 의논해봤는데요 FSK와 동일한 형태가 되도록 ASK를 만들어보려고 했습니다 그런데 그렇게 하려면 inputCode와 demodulateResult->getPayloadStr()를 비교해야 하는데, getPayloadStr은 Base 객체의 getBaseString의 결과를 return하더라구요 ㅠㅠ 다시보니까 FSKTest.cpp의 테스트 케이스도 모두 0~f 로 이루어져있더라구요. 그래서 0과 1로 구성된 문자열로만 테스트를 하게 되었습니다 😢

```cpp
TEST_P(ASKTestFixture, ASKModulationStringTest){
    createASK();

    string source;
    string expectedModulateResult;
    int expectedCodeLength;

    std::tie(source, expectedCodeLength) = GetParam();
    HexVector hexVector = ASCIICharset().encode(source);
    auto *packet = new Packet(BaseType::BASE2, hexVector);

    auto modulateASK = ask->modulate(packet->toString());
    EXPECT_EQ(modulateASK.size(), expectedCodeLength);

//    auto demodulateResult = ask->demodulate(modulateASK);
//    EXPECT_EQ(source, demodulateResult->getPayloadStr());
}
```
@kuro11pow2
아 위 테스트의 ASCIICharset 세팅은 @jopopcorn 님이 추가적으로 넣으신거였군요. ㅋㅋㅋ
ASK만의 단위 테스트이기 때문에 해당 테스트에서는 0과 1만 확인하는 것이 맞겠지만
여러분들이 작성한 utf-8과의 조합도 한번 보고 싶었는데 아쉽네용 ㅎㅎ

@kuro11pow2 @jopopcorn @psh2849 @HACore 
ASK 구현하느라 다들 너무너무 고생했어요!! :) 리뷰는 좀 더 꼼꼼하게 보고 드리도록 하겠습니다!
PR 제안만으로도 대단합니당! 굿굿!!

**Reply by HACore** [reply]
> 충분히 모호할 수 있습니다. **modulate과 demodulate는 정반대라서 깔끔하게 정반대가 되는 것이 맞는 것 아닌가요?** 라는 질문이죠?
> 
> `demodulate`는 rx 부분에서 사용되는 함수이고 `S`의 존재는 음파 인식 트리거를 위한 기준점이라고 볼 수 있습니다. AI Assistant로 예를 들자면 갤럭시는 하이 빅스비! Google Assistant는 Hey Google, 아이폰은 시리야 정도로 볼 수 있어요. 이러한 특정 트리거 이후에 음성 인식 모드에 진입하게 되죠.
> 
> 음파 통신은 `S`에 해당하는 (18002 - 86)hz 를 의미하고 이 주파수가 감지된 이후에 데이터를 인식합니다. 그 데이터가 `S` 다음에 오는 데이터들을 의미합니다. 위에 언급된 AI Assistant들처럼 음파 인식 단에서 listen 호출 이후로 지속적으로 `S`에 해당하는 주파수를 확인합니다. 발생하는지 안하는지 그리고 발생과 동시에 뒤에 발생하는 데이터들을 쭈우우욱 받아서 `demodulate`를 하게 될거에요. 따라서 Modem의 `demodulate`에서 `S`에 대한 해석을 하지 않게 되었답니다.
> 
> 근데 `modulate`에는 왜 `S`가 있는건가? 이것은 개발 용이성을 위해 이곳에서 구현하게 되었어요. 주파수를 oscillate 하는 Responsibility를 Wave에서만 주고 싶었던 것이죠.
> 
> 더 나은 개발 용이한 포인트를 찾게 된다면 나중에 바뀔수도 있습니다! 답변이 되었을까용?!

아하 모듈레이션에 너무 집중해서 음파를 인식하는 전체적인 매커니즘을 고려하지 못했었네요..!
찰떡같은 예시로 바로 이해가 됐어요ㅎㅎ 감사합니다!


> @HACore 지금에서야 확인해서 답변드려요! 질문 하나밖에 안했는데 죄송하다니요 ㅋㅋ (오랜만에 질문을 주셨어요 ㅠㅎㅎ)

관련 질문을 이미 앞에서 수없이 받으셨는데 더이상은 멘토님 괴롭히기(?)가 아닐까 걱정되서 말씀드린 부분이었습니당😂😂

**Comment by judemin** [comment]
고생 많으셨습니다!
